File: mat_ops/jax/mat_scale.py
import os, sys
import time

sys.path.append(os.path.dirname(os.path.dirname(__file__)))
import mat_bench

import jax
import jax.numpy as jnp

jax.config.update("jax_enable_x64", True)
dtype = jnp.dtype(jnp.float32)
# jax.config.update("jax_platform_name", "cpu")


@jax.jit
def kernel(x, a):
    print("Tracing Jax function ...............")
    return a * x


class mat_scale(mat_bench.mat_bench):

    def __init__(self):
        super().__init__()
        self.elem_size = dtype.itemsize
        self.memory_transfer_per_loop = (2.0 * float(self.elem_size) * 2.0 *
                                         float(self.total_size) /
                                         (1024.0 * 1024.0 * 1024.0))

    def benchmark(self):
        self.print_bench()
        print("\nSimulation info: 2d mat scale\n")

        x = jnp.ones((self.side_size, self.side_size), dtype=dtype)
        y = jnp.zeros((self.side_size, self.side_size), dtype=dtype)
        self.assert_shape_and_dtype(x)
        self.assert_shape_and_dtype(y)

        # warm up
        y = kernel(x, 2.0)
        y = kernel(y, 0.5)

        #main loop
        self.loops = 0
        start_time = time.perf_counter()
        while ((time.perf_counter() - start_time) < self.benchtime):
            y = kernel(x, 2.0)
            y = kernel(y, 0.5)
            self.loops = self.loops + 1
        y.block_until_ready()
        self.duration = time.perf_counter() - start_time

        self.test_result(y, float(self.total_size))
        self.print_performance()


if __name__ == "__main__":
    test = mat_scale()
    test.benchmark()

File: mat_ops/cpu/mat_scale.cpp
#include <chrono>
#include <string>
#include <omp.h>
#include <thread>
#include "array2d.h"
#include "mat_bench.h"

#define value_t float
#define index_t int

struct mat_scale : public mat_bench<value_t, index_t>
{

    void benchmark()
    {
        print_bench();

        std::cout << "\nSimulation info: 2d mat scale\n";

        value_t **x = create_array2d<value_t, index_t>(side_size, side_size);
        value_t doubleit = 2.0;
        value_t halfit = 0.5;

#pragma omp parallel for
        for (index_t i = 0; i < side_size; i++)
        {
            for (index_t j = 0; j < side_size; j++)
            {
                x[i][j] = 1.0;
            }
        }

        loops = 0;
        auto start = std::chrono::high_resolution_clock::now();
        while ((std::chrono::duration_cast<std::chrono::milliseconds>(
                    std::chrono::high_resolution_clock::now() - start)
                    .count()) < 1000.0 * benchtime)
        {
#pragma omp parallel for
            for (index_t i = 0; i < side_size; i++)
            {
                for (index_t j = 0; j < side_size; j++)
                {
                    x[i][j] *= doubleit;
                }
            }
#pragma omp parallel for
            for (index_t i = 0; i < side_size; i++)
            {
                for (index_t j = 0; j < side_size; j++)
                {
                    x[i][j] *= halfit;
                }
            }
            loops++;
        }
        auto stop = std::chrono::high_resolution_clock::now();
        duration = 1.0e-3 * std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count();

        test_result(x, value_t(total_size));
        print_performance();

        destroy_array2d<value_t, index_t>(x);
    }

    mat_scale(int narg, char **arg) : mat_bench<value_t, index_t>(narg, arg)
    {
        memory_transfer_per_loop = 2.0 * sizeof(value_t) * 2.0 * double(total_size) /
                                   (1024.0 * 1024.0 * 1024.0);
    }
};

int main(int narg, char **arg)
{
    mat_scale test(narg, arg);
    test.benchmark();
}